File: phone/ios/Sources/DashboardClient.swift
import Foundation

/// Bir veya birden fazla dashboard/v1 host'undan snapshot çeker.
/// Çok-makineli birleşik görünüm: maliyet toplanır, kota yüzdeleri toplanmaz (hesap bazlı; id tekil).
struct DashboardClient {
    struct Host { let baseUrl: String; let token: String? }

    func fetch(_ host: Host) async throws -> DashboardSnapshot {
        guard let url = URL(string: "\(host.baseUrl.trimmingTrailingSlash())/dashboard/v1/snapshot") else {
            throw URLError(.badURL)
        }
        var req = URLRequest(url: url)
        if let token = host.token { req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse, http.statusCode == 200 else {
            throw URLError(.userAuthenticationRequired)
        }
        return try JSONDecoder().decode(DashboardSnapshot.self, from: data)
    }

    func fetchMerged(_ hosts: [Host]) async -> [ProviderRow] {
        var byId: [String: ProviderRow] = [:]
        var order: [String] = []
        for host in hosts {
            guard let snap = try? await fetch(host) else { continue }
            for row in snap.providers {
                if var existing = byId[row.id] {
                    // Maliyet topla, kota pencerelerini tekil tut.
                    var cost = existing.cost ?? CostInfo()
                    cost.todayUSD = (cost.todayUSD ?? 0) + (row.cost?.todayUSD ?? 0)
                    cost.last30DaysUSD = (cost.last30DaysUSD ?? 0) + (row.cost?.last30DaysUSD ?? 0)
                    existing.cost = cost
                    byId[row.id] = existing
                } else {
                    byId[row.id] = row
                    order.append(row.id)
                }
            }
        }
        return order.compactMap { byId[$0] }.sorted { ($0.display?.sortKey ?? 0) < ($1.display?.sortKey ?? 0) }
    }
}

private extension String {
    func trimmingTrailingSlash() -> String { hasSuffix("/") ? String(dropLast()) : self }
}

File: phone/android/app/src/main/java/com/codexbridge/widget/DashboardClient.kt
package com.codexbridge.widget

import io.ktor.client.HttpClient
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.statement.bodyAsText
import io.ktor.http.HttpHeaders
import kotlinx.serialization.json.Json

/**
 * Bir veya birden fazla dashboard/v1 host'undan snapshot çeker.
 *
 * Çok-makineli birleşik görünüm: birden fazla host verilince snapshot'lar birleştirilir.
 * KRİTİK KURAL — maliyet toplanır, kota yüzdeleri toplanmaz (hesap bazlı; aynı id tekilleştirilir).
 */
class DashboardClient(
    private val http: HttpClient,
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    data class Host(val baseUrl: String, val token: String?)

    suspend fun fetch(host: Host): DashboardSnapshot {
        val text = http.get("${host.baseUrl.trimEnd('/')}/dashboard/v1/snapshot") {
            host.token?.let { header(HttpHeaders.Authorization, "Bearer $it") }
        }.bodyAsText()
        return json.decodeFromString(DashboardSnapshot.serializer(), text)
    }

    /** Birden fazla host'un satırlarını birleştirir: maliyet toplanır, kota (aynı id) tekil. */
    suspend fun fetchMerged(hosts: List<Host>): List<ProviderRow> {
        val byId = LinkedHashMap<String, ProviderRow>()
        for (host in hosts) {
            val snap = runCatching { fetch(host) }.getOrNull() ?: continue
            for (row in snap.providers) {
                val existing = byId[row.id]
                if (existing == null) {
                    byId[row.id] = row
                } else {
                    // Kota pencerelerini tekil tut (hesap bazlı), yalnızca maliyeti topla.
                    val mergedCost = CostInfo(
                        todayUsd = (existing.cost?.todayUsd ?: 0.0) + (row.cost?.todayUsd ?: 0.0),
                        last30DaysUsd = (existing.cost?.last30DaysUsd ?: 0.0) + (row.cost?.last30DaysUsd ?: 0.0),
                    )
                    byId[row.id] = existing.copy(cost = mergedCost)
                }
            }
        }
        return byId.values.sortedBy { it.display?.sortKey ?: 0 }
    }
}
